File: scripts/fetch_stocks.py
"""
台股資料抓取腳本
從 TWSE (上市) 和 TPEX (上櫃) 抓取所有公司基本資料
"""
import requests
import csv
import json
from io import StringIO

def fetch_twse_stocks():
    """抓取上市公司清單"""
    print("正在抓取上市公司資料...")
    url = "https://mopsfin.twse.com.tw/opendata/t187ap03_L.csv"
    
    try:
        response = requests.get(url, timeout=30)
        response.encoding = 'utf-8'
        
        # 解析 CSV
        csv_data = StringIO(response.text)
        reader = csv.DictReader(csv_data)
        
        stocks = []
        for row in reader:
            stock = {
                'id': row['公司代號'].strip(),
                'name': row['公司名稱'].strip(),
                'shortName': row['公司簡稱'].strip(),
                'industry': row['產業別'].strip(),
                'market': '上市',
                'description': '',  # 待補充
                'keyMetric': ''     # 待補充
            }
            stocks.append(stock)
        
        print(f"✓ 成功抓取 {len(stocks)} 家上市公司")
        return stocks
    
    except Exception as e:
        print(f"✗ 抓取上市公司失敗: {e}")
        return []

def fetch_tpex_stocks():
    """抓取上櫃公司清單"""
    print("正在抓取上櫃公司資料...")
    # TPEX 使用類似的 API
    url = "https://www.tpex.org.tw/openapi/v1/tpex_mainboard_daily_close_quotes"
    
    try:
        response = requests.get(url, timeout=30)
        data = response.json()
        
        stocks = []
        seen_ids = set()
        
        # 從當日行情中提取上櫃股票代號
        for item in data:
            stock_id = item.get('SecuritiesCompanyCode', '').strip()
            stock_name = item.get('CompanyName', '').strip()
            
            if stock_id and stock_id not in seen_ids:
                seen_ids.add(stock_id)
                stock = {
                    'id': stock_id,
                    'name': stock_name,
                    'shortName': stock_name,  # 上櫃資料較少，使用全名
                    'industry': '上櫃',  # 簡化處理
                    'market': '上櫃',
                    'description': '',
                    'keyMetric': ''
                }
                stocks.append(stock)
        
        print(f"✓ 成功抓取 {len(stocks)} 家上櫃公司")
        return stocks
    
    except Exception as e:
        print(f"✗ 抓取上櫃公司失敗: {e}")
        print("  將使用備用 CSV 來源...")
        # 備用方案：使用 CSV
        return fetch_tpex_from_csv()

def fetch_tpex_from_csv():
    """備用：從 CSV 抓取上櫃資料"""
    try:
        url = "https://www.tpex.org.tw/openapi/v1/tpex_mainboard_securities"
        response = requests.get(url, timeout=30)
        data = response.json()
        
        stocks = []
        for item in data:
            stock = {
                'id': item.get('SecuritiesCompanyCode', '').strip(),
                'name': item.get('CompanyName', '').strip(),
                'shortName': item.get('CompanyName', '').strip(),
                'industry': '上櫃',
                'market': '上櫃',
                'description': '',
                'keyMetric': ''
            }
            if stock['id']:
                stocks.append(stock)
        
        return stocks
    except:
        return []

def merge_and_save(twse_stocks, tpex_stocks, output_file='stocks_raw.json'):
    """合併資料並儲存"""
    all_stocks = twse_stocks + tpex_stocks
    
    # 按代號排序
    all_stocks.sort(key=lambda x: x['id'])
    
    print(f"\n總共 {len(all_stocks)} 家公司")
    print(f"儲存至 {output_file}...")
    
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(all_stocks, f, ensure_ascii=False, indent=2)
    
    print(f"✓ 資料已儲存")
    return all_stocks

if __name__ == '__main__':
    print("=" * 50)
    print("台股資料抓取工具")
    print("=" * 50)
    
    # 抓取資料
    twse = fetch_twse_stocks()
    tpex = fetch_tpex_stocks()
    
    # 合併並儲存
    if twse or tpex:
        all_stocks = merge_and_save(twse, tpex)
        print(f"\n✓ 完成！共 {len(all_stocks)} 筆資料")
    else:
        print("\n✗ 無法抓取任何資料")

File: scripts/fetch_stocks.js
/**
 * 台股資料抓取腳本 (Node.js)
 * 從 TWSE (上市) 和 TPEX (上櫃) 抓取所有公司基本資料
 */

const https = require('https');
const fs = require('fs');

// 抓取 TWSE 上市資料
async function fetchTWSEStocks() {
    console.log('正在抓取上市公司資料...');

    return new Promise((resolve, reject) => {
        const url = 'https://mopsfin.twse.com.tw/opendata/t187ap03_L.csv';

        https.get(url, (res) => {
            let data = '';

            res.on('data', (chunk) => {
                data += chunk;
            });

            res.on('end', () => {
                try {
                    // 解析 CSV
                    const lines = data.split('\n');
                    const headers = lines[0].split(',').map(h => h.replace(/"/g, '').trim());

                    const stocks = [];
                    for (let i = 1; i < lines.length; i++) {
                        if (!lines[i].trim()) continue;

                        // 簡化處理：處理引號包裹的欄位
                        const matches = lines[i].match(/("([^"]*)"|[^,]+)/g);
                        if (!matches || matches.length < 3) continue;

                        const values = matches.map(v => v.replace(/^"|"$/g, '').trim());

                        // 找到關鍵欄位的索引
                        const idIndex = headers.indexOf('公司代號');
                        const nameIndex = headers.indexOf('公司名稱');
                        const shortNameIndex = headers.indexOf('公司簡稱');
                        const industryIndex = headers.indexOf('產業別');

                        if (idIndex >= 0 && values[idIndex]) {
                            stocks.push({
                                id: values[idIndex],
                                name: values[nameIndex] || values[idIndex],
                                shortName: values[shortNameIndex] || values[nameIndex] || values[idIndex],
                                industry: values[industryIndex] || '上市',
                                market: '上市',
                                description: '',
                                keyMetric: ''
                            });
                        }
                    }

                    console.log(`✓ 成功抓取 ${stocks.length} 家上市公司`);
                    resolve(stocks);
                } catch (err) {
                    console.error('✗ 解析上市資料失敗:', err.message);
                    resolve([]);
                }
            });
        }).on('error', (err) => {
            console.error('✗ 抓取上市公司失敗:', err.message);
            resolve([]);
        });
    });
}

// 抓取 TPEX 上櫃資料
async function fetchTPEXStocks() {
    console.log('正在抓取上櫃公司資料...');

    return new Promise((resolve) => {
        const url = 'https://www.tpex.org.tw/openapi/v1/tpex_mainboard_securities';

        https.get(url, (res) => {
            let data = '';

            res.on('data', (chunk) => {
                data += chunk;
            });

            res.on('end', () => {
                try {
                    const json = JSON.parse(data);
                    const stocks = [];

                    for (const item of json) {
                        const id = item.SecuritiesCompanyCode || item.Code;
                        const name = item.CompanyName || item.Name;

                        if (id && id.trim()) {
                            stocks.push({
                                id: id.trim(),
                                name: name ? name.trim() : id,
                                shortName: name ? name.trim() : id,
                                industry: '上櫃',
                                market: '上櫃',
                                description: '',
                                keyMetric: ''
                            });
                        }
                    }

                    console.log(`✓ 成功抓取 ${stocks.length} 家上櫃公司`);
                    resolve(stocks);
                } catch (err) {
                    console.error('✗ 解析上櫃資料失敗:', err.message);
                    resolve([]);
                }
            });
        }).on('error', (err) => {
            console.error('✗ 抓取上櫃公司失敗:', err.message);
            resolve([]);
        });
    });
}

// 主函數
async function main() {
    console.log('='.repeat(50));
    console.log('台股資料抓取工具 (Node.js)');
    console.log('='.repeat(50));

    // 抓取資料
    const twse = await fetchTWSEStocks();
    const tpex = await fetchTPEXStocks();

    // 合併
    const allStocks = [...twse, ...tpex];

    // 排序
    allStocks.sort((a, b) => a.id.localeCompare(b.id));

    console.log(`\n總共 ${allStocks.length} 家公司`);
    console.log('儲存至 stocks_raw.json...');

    // 儲存
    fs.writeFileSync(
        'stocks_raw.json',
        JSON.stringify(allStocks, null, 2),
        'utf-8'
    );

    console.log(`✓ 完成！共 ${allStocks.length} 筆資料`);

    // 顯示前 5 筆作為範例
    console.log('\n前 5 筆資料範例:');
    console.log(JSON.stringify(allStocks.slice(0, 5), null, 2));
}

main().catch(console.error);
